Upstream Unicam driver

6by9 opened the pull request:
Whilst not merged, this takes the upstream Unicam driver, backports it to 5.4, and includes any of our fixes since then, and those raised by upstream. With those patches squashed it should be a reasonable v3.
It is dependent on another upstream patchset that hasn't been merged yet for dynamically changing the number of active CSI data lanes being used, so that has been backported as well.

DT has changed slightly, so those are all fixed up as well.

naushir committed:
media: bcm2835-unicam: Driver for CCP2/CSI2 camera interface

Add a driver for the Unicam camera receiver block on BCM283x processors.
Compared to the bcm2835-camera driver present in staging, this driver
handles the Unicam block only (CSI-2 receiver), and doesn't depend on
the VC4 firmware running on the VPU.

The commit is made up of a series of changes cherry-picked from the
rpi-5.4.y branch of https://github.com/raspberrypi/linux/ with
additional enhancements, forward-ported to the mainline kernel.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Naushir Patuck <naush@raspberrypi.com>
Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Reported-by: kbuild test robot <lkp@intel.com>

Commit:
media: bcm2835: unicam: Fix uninitialized warning

Signed-off-by: Jacko Dirks <jdirks.linuxdev@gmail.com>

6by9 committed:
media: bcm2835-unicam: Always service interrupts

From when bringing up the driver, there was a check in the isr
to ignore interrupts (claiming them handled) should the driver
not be streaming.

The VPU now will not register a camera driver if it finds a
CSI2 node enabled in device tree, therefore this flawed check is
redundant.

https://github.com/raspberrypi/linux/issues/3602

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: bcm2835-unicam: Retain packing information on G_FMT

The change to retrieve the pixel format always on g_fmt didn't
check whether the native or unpacked version of the format
had been requested, and always returned the packed one.
Correct this so that the packing setting is retained whereever
possible.

Fixes "9d59e89 media: bcm2835-unicam: Re-fetch mbus code from subdev
on a g_fmt call"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

davidplowman committed:
media: bcm2835-unicam: change minimum number of vb2_queue buffers to 1

Since the unicam driver was modified to write to a dummy buffer when no
user-supplied buffer is available, it can now write to and return a
buffer even when there's only a single one. Enable this by changing the
min_buffers_needed in the vb2_queue; it will be useful for enabling
still captures without allocating more memory than absolutely necessary.

Signed-off-by: David Plowman <david.plowman@raspberrypi.com>

Commit:
media: v4l2-subdev: Introduce [get|set]_mbus_config pad ops

Upstream https://patchwork.linuxtv.org/patch/64669/

Introduce two new pad operations to allow retrieving and configuring the
media bus parameters on a subdevice pad.

The newly introduced operations aims to replace the s/g_mbus_config video
operations, which have been on their way for deprecation since a long
time.

Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

Commit:
media: i2c: Use the new get_mbus_config pad op

Upstream https://patchwork.linuxtv.org/patch/64669/

Move the existing users of the g_mbus_config video operation to use the
newly introduced get_mbus_config pad operations.

All the ported drivers report a static media bus configuration and do no
support s_mbus_config so the operation implementation has not changed.

Bridge drivers needs to call the new pad operation and will receive an
-ENOICTLCMD when calling the old g_mbus_config video operation

Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

Commit:
media: i2c: ov6650: Use new [get|set]_mbus_config ops

Upstream https://patchwork.linuxtv.org/patch/64674/

Use the new get_mbus_config and set_mbus_config pad operations in place
of the video operations currently in use.

Compared to other drivers where the same conversion has been performed,
ov6650 proved to be a bit more tricky, as the existing g_mbus_config
implementation did not report the currently applied configuration but
the set of all possible configuration options.

Adapt the driver to support the semantic of the two newly introduced
operations:
- get_mbus_config reports the current media bus configuration
- set_mbus_config applies only changes explicitly requested and updates
  the provided cfg parameter to report what has actually been applied to
  the hardware.

Compile-tested only.

Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

Commit:
media: pxa_camera: Use the new set_mbus_config op

Upstream https://patchwork.linuxtv.org/patch/64671/

Move the PXA camera driver to use the new set_mbus_config pad operation.
For this platform the change is not only cosmetic, as the pxa driver is
currently the only driver in mainline to make use of the g_mbus_config
and s_mbus_config video operations.

The existing driver semantic is the following:
- Collect all supported mbus config flags from the remote end
- Match them with the supported PXA mbus configuration flags
- If the remote subdevice allows multiple options for for VSYNC, HSYNC
  and PCLK polarity, use platform data requested settings

The semantic of the new get_mbus_config and set_mbus_config differs from
the corresponding video ops, particularly in the fact get_mbus_config
reports the current mbus configuration and not the set of supported
configuration options, with set_mbus_config always reporting the actual
mbus configuration applied to the remote subdevice.

Adapt the driver to perform the following
- Set the remote subdevice mbus configuration according to the PXA
  platform data preferences.
- If the applied configuration differs from the requested one (i.e. the
  remote subdevice does not allow changing one setting) make sure that
  - The remote end does not claim for DATA_ACTIVE_LOW, which seems not
    supported by the platform
  - The bus mastering roles match

While at there remove a few checks performed on the media bus
configuration at get_format() time as they do not belong there.

Compile-tested only.

Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

Commit:
media: v4l2-subdev: Remove [s|g]_mbus_config video ops

Upstream https://patchwork.linuxtv.org/patch/64670/

With all sensor and platform drivers now converted to use the new
get_mbus_config and set_mbus_config pad operations, remove the
deprecated video operations g_mbus_config and s_mbus_config.

Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

Commit:
staging: media: imx: Update TODO entry

Upstream https://patchwork.linuxtv.org/patch/64672/

Update the TODO entry that mentioned a potential use case for the now
removed g_mbus_config video operation.

Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

Commit:
media: i2c: adv748x: Adjust TXA data lanes number

Upstream https://patchwork.linuxtv.org/patch/64673/

When outputting SD-Core output through the TXA MIPI CSI-2 interface,
the number of enabled data lanes should be reduced in order to guarantee
that the two video formats produced by the SD-Core (480i and 576i)
generate a MIPI CSI-2 link clock frequency compatible with the MIPI D-PHY
specifications.

Limit the number of enabled data lanes to 2, which is guaranteed to
support 480i and 576i formats.

Cache the number of enabled data lanes to be able to report it through
the new get_mbus_config operation.

Reviewed-by: Kieran Bingham <kieran.bingham+renesas@ideasonboard.com>
Reviewed-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Reviewed-by: Niklas Söderlund <niklas.soderlund+renesas@ragnatech.se>
Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

Commit:
media: i2c: adv748x: Implement get_mbus_config

Upstream https://patchwork.linuxtv.org/patch/64676/

Implement the newly introduced get_mbus_config operation to report the
number of currently used data lanes on the MIPI CSI-2 interface.

Reviewed-by: Kieran Bingham <kieran.bingham+renesas@ideasonboard.com>
Reviewed-by: Niklas Söderlund <niklas.soderlund+renesas@ragnatech.se>
Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

Commit:
media: rcar-csi2: Negotiate data lanes number

Upstream https://patchwork.linuxtv.org/patch/64675/

Use the newly introduced get_mbus_config() subdevice pad operation to
retrieve the remote subdevice MIPI CSI-2 bus configuration and configure
the number of active data lanes accordingly.

In order to be able to call the remote subdevice operation cache the
index of the remote pad connected to the single CSI-2 input port.

Signed-off-by: Jacopo Mondi <jacopo+renesas@jmondi.org>

6by9 committed:
drivers: media: Remove the downstream version of bcm2835-unicam

About to be replaced by the upstream version.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
include: media: Add vfl_devnode_type of VFL_TYPE_VIDEO

Upsstream are renaming VFL_TYPE_GRABBER to VFL_TYPE_VIDEO.
To make backporting the upstream Unicam driver easier, add an
extra enum entry (same as VFL_TYPE_GRABBER) to match that.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: bcm2835-unicam: Add support for get_mbus_config to set num lanes

Use the get_mbus_config pad subdev call to allow a source to use
fewer than the number of CSI2 lanes defined in device tree.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: bcm2835-unicam: Fixup review comments from Hans.

Updates the driver based on the upstream review comments from
Hans Verkuil at https://patchwork.linuxtv.org/patch/63531/

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dt/dtoverlays: Fix up base DT and overlays for updated Unicam driver

The upstreamed Unicam driver uses a dt property to denote how many
lanes are supported by the receiver peripheral, independent of
the number of lanes that the sensor wants to use. It also doesn't
check the remote endpoint config for the number of lanes as that
isn't the accepted way of doing things.

Update the base DT for the brcm,num-data-lanes property, and the
overlays to define the desired number of lanes at both ends of
the link.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
@davidplowman @naushir Please test. It's working for me with TC358743, but I haven't tried the camera modules and libcamera.

@pinchartl Hopefully a good next step for v3.
Drop caa8557f427a1f6c18fbbb6764ec827b6a2d4a30 and you could still upstream it without the dependency on Jacopo's patchset.

Now on to the ISP driver and vc-sm-cma....

naushir commented:
Looks to be working fine for me in libcamera.

naushir commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -2428,35 +2411,33 @@ static int register_node(struct unicam_device *unicam, struct unicam_node *node,
-	}
-	if (node->pad_id == METADATA_PAD ||
+	/* TODO: review which of these disables are still needed */
+	if (pad_id == METADATA_PAD ||
```
Is this TODO still needed?

6by9 replied:
Yes.
Hans' comment on https://patchwork.linuxtv.org/patch/63531/

> Some of these are probably no longer needed since the v4l2 core will disable them for you.

I've removed the disables of the event ioctls which we discussed the other day, but I didn't look into these.
I believe it is true that the core will disable some ioctls based on this being a metadata type, but it'd need careful checking, and may be different between 5.4 and linux-media-next.

pelwell commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -0,0 +1,2825 @@
+static int of_unicam_connect_subdevs(struct unicam_device *dev)
+{
+	struct platform_device *pdev = dev->pdev;
+	struct v4l2_fwnode_endpoint ep = { 0 };
```
I get a build warning on this line:
```
drivers/media/platform/bcm2835/bcm2835-unicam.c: In function 'of_unicam_connect_subdevs':
drivers/media/platform/bcm2835/bcm2835-unicam.c:2582:9: warning: missing braces around initializer [-Wmissing-braces]
  struct v4l2_fwnode_endpoint ep = { 0 };
         ^
drivers/media/platform/bcm2835/bcm2835-unicam.c:2582:9: warning: (near initialization for 'ep.base') [-Wmissing-braces]
```

pelwell replied:
It's happy with ` = { { 0 } };`.

popcornmix replied:
Isn't `= { }` the correct initialiser?
Including the `0` means the first element of structure needs to be assignable to zero which isn't the case with many types (including sub-structures)

pelwell replied:
I think the problem is that the first element of a `v4l2_fwnode_endpoint` is another struct - `fwnode_endpoint` - which has a scalar value as the first element.

popcornmix replied:
Yes but doesn't `= {}` avoid the warning?
It's a gcc construct, but is widely used in the kernel.

pelwell replied:
Are you saying you don't get a warning?

popcornmix replied:
I don't with `= {}` I do with `={0}`

pelwell replied:
```
phil@buildbot:~/pi/linux$ git grep '= { { 0 } };' |  wc
     54     533    5135
phil@buildbot:~/pi/linux$ git grep '= { { { 0 } } };' |  wc
      4      49     396
```

popcornmix replied:
```
$ git grep '= { };' | wc
    573    4043   41864
```

JamesH65 replied:
= {0} is the correct syntax to zero an entire structure to zero, but there is/was a known bug in gcc that meant it failed to compile in some cases. I thought it had been fixed though. I was on the distribution list for the fix because I reported it at some point in the ancient past.

popcornmix replied:
```
$ git grep '= {};' | wc
   1754   10450  132323
```

pelwell replied:
```
phil@buildbot:~/pi/linux$ git grep '= {};' |  wc
   1754   10450  132323
```

pelwell replied:
You win. It's not an extension I'd noticed before, but it makes more sense than the alternative.

6by9 replied:
It comes from the version of the patch that was being upstreamed, and they tend to use a later gcc than we appear to.
{0} is meant to zero the whole struct.

pelwell replied:
This isn't going to make for a particularly tricky merge conflict if it does change upstream.

6by9 replied:
I'm happy to create an additional downstream patch here to use memset instead to avoid the build warning, but I'm hoping Laurent will otherwise pick up the patches to create the next patchset to send to the list, and they'll want ```= { 0 }```

6by9 commented:
Rebased. Now just checking the disable_ioctl calls for metadata nodes.

pelwell commented:
Were you expecting to lose "media: bcm2835-unicam: Retain packing information on G_FMT" and "media: bcm2835-unicam: change minimum number of vb2_queue buffers to 1" around the time of the rebase?

6by9 commented:
> Were you expecting to lose "media: bcm2835-unicam: Retain packing information on G_FMT" and "media: bcm2835-unicam: change minimum number of vb2_queue buffers to 1" around the time of the rebase?

No :-(
Time to retrieve them.

6by9 committed:
media: bcm2835-unicam: Avoid gcc warning over {0} on endpoint

Older gcc versions object to = { 0 } initialisation if the first
elemtn in the structure is a substructure.

Use = { } to avoid this compiler warning.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
OK, retrieved the two missing commits, and added one to use memset instead of ={0}.
Apologies if I've introduced a build error here - my laptop is creaking in the heat, and probably wants a reboot too.

pelwell commented:
Did you not want to use ` = {};` (note the lack of a 0), since it is the most common way to initialise a structure to zero in the kernel?
Otherwise this looks good, and there are no build errors (or new warnings).

6by9 commented:
Up to you over ```= {};``` vs ```memset```. This is a downstream only patch so makes little difference in the grand scheme of things.

Give me a chance to check up on those disable_ioctl calls as it'd be nice to merge all changes in this one PR.

Anything else from @naushir?

pelwell commented:
I'd prefer `= {};` because it's one fewer line to read and one fewer line to contain mistakes.

I'll wait for the nod re: merging.

6by9 commented:
Updated with ``` = { };```

We do need the disable_ioctl calls as both video and metadata nodes are type VFL_TYPE_[VIDEO|GRABBER] (it got renamed), and the core disabling of ioctls is based on that type. We want to disable them based on whether the source subdev supports the relevant subdev API call or not, so our struct of ops always contains the superset of all of them.

6by9 commented:
I'm all happy now.

pelwell commented:
Just in time for the weekend.
